libsplinter/src/auth/oauth/builder/mod.rs

```diff
@@ -0,0 +1,169 @@
+    pub fn build(self) -> Result<OAuthClient, OAuthClientBuildError> {
+        let client_id = self.client_id.ok_or_else(|| {
+            InvalidStateError::with_message(
+                "A client ID is required to successfully build an OAuthClient".into(),
+            )
+        })?;
+        let client_secret = self.client_secret.ok_or_else(|| {
+            InvalidStateError::with_message(
+                "A client secret is required to successfully build an OAuthClient".into(),
+            )
+        })?;
+        let auth_url = self.auth_url.ok_or_else(|| {
+            InvalidStateError::with_message(
+                "An auth URL is required to successfully build an OAuthClient".into(),
+            )
+        })?;
+        let redirect_url = self.redirect_url.ok_or_else(|| {
+            InvalidStateError::with_message(
+                "A redirect URL is required to successfully build an OAuthClient".into(),
+            )
+        })?;
+        let token_url = self.token_url.ok_or_else(|| {
+            InvalidStateError::with_message(
+                "A token URL is required to successfully build an OAuthClient".into(),
+            )
+        })?;
+        let identity_provider = self.identity_provider.ok_or_else(|| {
+            InvalidStateError::with_message(
+                "An identity provider is required to successfully build an OAuthClient".into(),
+            )
+        })?;
+        Ok(OAuthClient::new(
+            client_id,
+            client_secret,
+            auth_url,
+            redirect_url,
+            token_url,
+            self.scopes,
+            identity_provider,
+        )?)
+    }
+
+    /// Sets the client ID for the OAuth2 provider.
+    pub fn with_client_id(mut self, client_id: String) -> Self {
+        self.client_id = Some(client_id);
+        self
+    }
+
+    /// Sets the client secret for the OAuth2 provider.
+    pub fn with_client_secret(mut self, client_secret: String) -> Self {
+        self.client_secret = Some(client_secret);
+        self
+    }
+
+    /// Sets the authorize URL for the OAuth2 provider.
+    pub fn with_auth_url(mut self, auth_url: String) -> Self {
+        self.auth_url = Some(auth_url);
+        self
+    }
+
+    /// Sets the redirect URL for the OAuth2 provider.
+    pub fn with_redirect_url(mut self, redirect_url: String) -> Self {
+        self.redirect_url = Some(redirect_url);
+        self
+    }
+
+    /// Sets the token URL for the OAuth2 provider.
+    pub fn with_token_url(mut self, token_url: String) -> Self {
+        self.token_url = Some(token_url);
+        self
+    }
+
+    /// Sets the scopes to request from the OAuth2 provider.
+    pub fn with_scopes(mut self, scopes: Vec<String>) -> Self {
+        let mut scopes = scopes;
+        self.scopes.append(&mut scopes);
+        self
+    }
+
+    /// Sets the identity provider to use to request a provider user identity reference.
```

Comment:
"provider user identity reference" is confusing. Maybe just "Sets the identity provider to use to request the user's identity from the OAuth2 provider"?

Reply:
Fixed